core/services/vrf/listener_v2.go

```diff
@@ -321,141 +377,250 @@ func (lsn *listenerV2) processRequestsPerSub(
+			// This shouldn't happen
+			l.Errorw("Error creating getCommitment payload. Proceeding anyway.",
+				"err", err)
+			return fulfilled
+		}
+
+		reqBlockNumber := new(big.Int).SetUint64(req.req.Raw.BlockNumber)
+		currBlock := new(big.Int).SetUint64(lsn.getLatestHead())
+		m := bigmath.Max(reqBlockNumber, currBlock)
+
+		var result string
+		calls[i] = rpc.BatchElem{
+			Method: "eth_call",
+			Args: []interface{}{
+				map[string]interface{}{
+					"to":   lsn.coordinator.Address(),
+					"data": hexutil.Bytes(payload),
+				},
+				// The block at which we want to make the call
+				hexutil.EncodeBig(m),
+			},
+			Result: &result,
+		}
+	}
+
+	err := lsn.ethClient.BatchCallContext(context.Background(), calls)
+	if err != nil {
+		l.Errorw("Unable to check if already fulfilled, failed batch call, processing anyways",
+			"err", err)
+		return fulfilled
+	}
+
+	for i, call := range calls {
+		if call.Error != nil {
+			l.Errorw("Unable to check if already fulfilled, processing anyways",
+				"reqID", reqs[i].req.RequestId.String(),
+				"txHash", reqs[i].req.Raw.TxHash,
+				"err", call.Error)
 			continue
 		}
 
-		// If we successfully enqueued for the txm, subtract that balance
-		// And loop to attempt to enqueue another fulfillment
-		startBalanceNoReserveLink.Sub(startBalanceNoReserveLink, maxLink)
-		processed[vrfRequest.RequestId.String()] = struct{}{}
-		incProcessedReqs(lsn.job.Name.ValueOrZero(), lsn.job.ExternalJobID, v2)
+		rString, ok := call.Result.(*string)
+		if !ok {
+			l.Errorw("Unexpected result from batch call",
+				"result", call.Result,
+				"reqID", reqs[i].req.RequestId.String(),
+				"txHash", reqs[i].req.Raw.TxHash)
+		}
+		result, err := hexutil.Decode(*rString)
+		if err != nil {
+			l.Errorw("Error decoding batch call result",
+				"result", call.Result,
+				"resultString", rString,
+				"reqID", reqs[i].req.RequestId.String(),
+				"txHash", reqs[i].req.Raw.TxHash)
+		}
+
+		if utils.IsEmpty(result) {
+			l.Infow("Request already fulfilled",
+				"reqID", reqs[i].req.RequestId.String(),
+				"txHash", reqs[i].req.Raw.TxHash)
+			fulfilled[i] = true
+		}
 	}
 
-	lggr.Infow("Finished processing for sub",
-		"endBalance", startBalanceNoReserveLink,
-		"total reqs", len(reqs),
-		"total processed", len(processed),
-		"total unique", uniqueReqs(reqs),
-		"time", time.Since(start).String())
-	return processed
+	l.Debugw("Done checking fulfillment status",
+		"numChecked", len(reqs), "time", time.Since(start).String())
+	return fulfilled
+}
+
+func (lsn *listenerV2) runPipelines(
+	l logger.Logger,
+	maxGasPriceWei *big.Int,
+	reqs []pendingRequest,
+) []vrfPipelineResult {
+	var (
+		start   = time.Now()
+		results = make([]vrfPipelineResult, len(reqs))
+		wg      = sync.WaitGroup{}
+	)
+
+	for i, req := range reqs {
+		wg.Add(1)
+		go func(i int, req pendingRequest) {
+			defer wg.Done()
+			results[i] = lsn.getMaxLinkForFulfillment(maxGasPriceWei, req, l)
```

Comment:
Waitgroups make me nervous about delaying shutdown. I noticed a `context.Background()` in `getMaxLinkForFulfillment` too. Maybe would be a good idea to construct a `ctx` from `lsn.ChStop` or whatever to pass in here and wire through wherever possible?

Reply:
Yeah good call, done. I created a context based on chStop in the request processing loop and passed it in everywhere.